### ISSUE TITLE
Implement --snapshot-only command line option (counterpart to --resume-only)

### ISSUE THREAD

**Issue by FlorianFranzen** [digint/btrbk#150]
I want to suggest an enhancement: Two new commands that allow you to only run run the snapshot or backup functionality individually.

To achieve this at the moment one has to create two config files, one for snapshots by removing all ```target ...``` lines and another one for backups with ```snapshot_create no``` and ```snapshot_preserve_min   all``` set.

Having two config files however makes everything more complex, including the use of the ```dryrun``` or ```list ...``` functionality and in the end it is easiest to just create a third config file by merging the other two. Now whenever you want to change anything, you will have to do so in up to three config files.

The motivation behind running the step individually is that snapshots might require certain services to be stopped during the taking of the snapshot but keeping them stopped during the potentially long lasting backup step creates unnecessary downtime and separating the two steps solves that problem. 

Any feedback on the idea is appreciated.

**Comment by digint** [comment]
At least one part of what you are suggesting is already implemented: if you run `btrbk --resume-only`, no snapshots are generated.

I never implemented the the counterpart (which would be called something like `btrbk --snapshot-only`, because I did not see a real use-case for this. Having a single config is probably such a use-case.

If you are talking of remote machines, I recommend to have a separate config on the remote which does only snapshotting, and a config on the "backup-server" for fetching the snapshots (having `snapshot_create no`). This way you make sure that snapshots are always created, even when network is down.

**Comment by FlorianFranzen** [comment]
Thank you for the quick feedback. For now it is all on one host.

I started working on what I assumed would be the simplest way to implement what I am looking on my fork [here](https://github.com/FlorianFranzen/btrbk/commit/81adca952f8c98df536af7987beebafb0033ed68).

I am not sure if that is the way to do it, so I wanted to get some feedback first, before I start testing any of these changes. My change would, in contrast of what you suggested, add new commands to btrbk instead of just flags.

**Comment by digint** [comment]
I'm afraid I don't have time to look at it in detail, but here's my first feedback on this:

1. please don't introduce new actions, they will clash with `dryrun` (which I admit is silly/redundant with `-n`, but it's there...). Instead, introduce `-s, --snapshot-only`, analogue to `--resume-only` (which sets `$resume_only`)
2. the rest is hard for me to check on the github diff, but I guess it's pretty straight-forward, simply disable all/some stuff in `# create backups`, which it seems you already did.

**Comment by FlorianFranzen** [comment]
Ok, that sound reasonable.

One last discussion point is how to trigger the rotation and therefore deletion of old snapshots (and their backups), when only running btrbk with the  ```-s``` and ```-r``` flag.

If I implement the ```-s``` flag like the ```-r``` flag the deletion code is never run, which I think is a problem.

Alternatively the ```-s``` flag could also runs deletion of snapshots and we introduce a third flag ```-b, --backup-only``` that acts like ```-r``` but also deletes old backups.

Any feedback is appreciated.

**Comment by digint** [comment]
I think both `--resume-only` and `--snapshot-only` should actually remove backups / snapshots. Now a quick look at the code shows me that `--resume-only` skips the backup removing part, I actually don't remember why I implemented it like this (probably because of the naming "resume" which kind of implies no deletion).

We already have a "do not delete things" flag: `-p, --preserve`, so if you want to snapshot without deletion, you would combine `--snapshot-only --preserve`.

#### Summing up, my proposal is to:

1. put deprecation flag on `--resume-only`
2. implement `-b, --backup-only` which does the same as `--resume-only`, but without skipping backup deletion (unless `-p` option is set)
3. implement `-s, --snapshot-only`, which skips the send/receive part but still deletes snapshots (unless `-p` option is set)
4. make sure that both options **never delete latest common subvolumes**


And maybe find another naming for the options, as I think the name "backup-only" could be misleading (because `btrbk run` does "backups", and for some setups the snapshots can also be regarded as backups). But actually I don't have many ideas right now (maybe "--target-only", this would match with "target_preserve" naming).

**Comment by digint** [comment]
I implemented the proposal above, in the [snapshot_only branch](/digint/btrbk/tree/snapshot_only).

First tests show that it works fine, but I fear that the naming could be a bit confusing. `snapshot-only` somehow reads "btrbk will only create snapshots", but the implementation still deletes snapshots according to the snapshot_preserve policy (unless `-p, --preserve` option is set). I can't think of a better naming for this right now though...

Any input is appreciated!

**Comment by asssaf** [comment]
I think snapshot-only makes sense.

How about --no-send or --disable-send?
Another option would be to enhance the config file to be able to skip targets if they're not mounted

**Comment by DL6AKU** [comment]
```--local-only``` or ```--local``` ?

**Comment by digint** [comment]
> How about --no-send or --disable-send?
> --local-only or --local ?

Nope, as dependent on your configuration both snapshots/backups can also be either remote or local...
Guess I'll keep it `snapshot-only`, `backup-only` and merge it before I forget about this branch.

**Comment by FlorianFranzen** [comment]
@digint Wow, thank you for implementing this so quickly. Is their a way to sponsor you some beverages?

Looks good to me. The naming of the options, together with the changes you made to the documentations, should make their functionality pretty clear.

**Comment by digint** [comment]
@FlorianFranzen sure, I just added a donate link [in README.md](/digint/btrbk#donate). Hope it works, and thanks in advance!

**Comment by digint** [comment]
I just figured that my proposal/implementation does not cover a common use case: two cron jobs, one for snapshots only and one for backup:

1. periodical snapshot, without deleting anything: would be `--snapshot-only --preserve`
2. periodical backup, which also deletes snapshots on success: something like `--do-backup --prune-backups --prune-snapshots`

**Comment by FlorianFranzen** [comment]
l think the necessity to delete snapshots only after they have been backed up depends on the relation of snapshot and backup spacing to the cron's frequency.

It should only be necessary to keep old snapshots if the time between successful backups is larger than the smallest time spacing between your snapshots times the number of snapshots (Assuming you are not trying to backup more frequently then you snapshot, which is nonsensical anyway).

For example, lets consider the following
```
snapshot_preserve_min   no
snapshot_preserve          7d 4w 6m

target_preserve_min     no
target_preserve            30d 8w *m
```
In that case the snapshot cron has to be run daily. The backup cron has to run (and succeed) at least once a week, else daily backup will be lost. 

If you run the snapshot and backup command after each other at the normal required frequency and react to backup failures before the snapshots that were not backed up get deleted again (in my example within a week), no snapshots should be lost.

Not to mention, that you have the same issue if you run btrbk with separate snapshot and backup configs.

I think we should be fine, as long as we warn users about the implications of running snapshots and backups separately, together with some general remarks about the effect of cron spacing and it relation to the preservation settings in the config.

I think implementing backup-aware snapshot deletion should be a separate feature of itself.

**Comment by asssaf** [comment]
The backup could be run manually (for example, if an external disk needs to be attached). Can't rely on an automated job running before snapshots are deleted.

A --snapshot-only --preserve is a needed use case so you can have a cron job for local-only snapshotting that won't fail (like it does now) when the external disk isn't connected, or at least add an option that causes the command to still return success even if backup wasn't done.

But maybe it's a separate use case.

**Comment by digint** [comment]
I implemented a new approach for this, which is (in my opinion) much less confusing as the previous approach: [snapshot_resume_commands branch](/digint/btrbk/tree/snapshot_resume_commands)

Instead of doing negative logic (--snapshot-only, --backup-only), I added new commands `btrbk snapshot` and `btrbk resume`, as well as `--preserve-snapshots` and `--preserve-backups` options.

I think all the use cases should be covered this way:

|Command|Result|
|:----|:----|
|`btrbk run`| create snapshots, create backups, delete backups, delete snapshots|
|`btrbk run --preserve`| create snapshots, create backups|
|`btrbk run --preserve-snapshots`| create snapshots, create backups, delete backups|
|`btrbk run --preserve-backups`| create snapshots, create backups, delete snapshots|
|`btrbk snapshot`| create snapshots, delete snapshots|
|`btrbk snapshot --preserve`|create snapshots|
|`btrbk resume`|create backup, delete backups, delete snapshots|
|`btrbk resume --preserve`|create backup|
|`btrbk resume --preserve-snapshots`|create backup, delete backups|
|`btrbk resume --preserve-backups`|create backup, delete snapshots|

**Comment by asssaf** [comment]
So for my use case i will need to run btrbk twice:
1. btrbk snapshot --preserve
2. btrbk resume (if the external drive is connected)

Looks good to me!

**Comment by FlorianFranzen** [comment]
Indeed, that covers it all. Great work.

I would add ```backup``` as an alias for ```resume```, but that is more personal taste and of course up to you. 

Thanks again for sharing your great work.

**Comment by digint** [comment]
implemented (squashed commit) in: 5428e9cd93f24cc4ecd77651db876d2f82a70adc
